boards/BMS/Src/Io/Io_LTC6813.c

```diff
@@ -0,0 +1,374 @@
+ * to check the status of ADC conversions was not transmitted or received
+ * successfully.
+ */
+static ExitCode Io_LTC6813_PollAdcConversion(void);
+
+static void Io_LTC6813_ParseCellsAndPerformPec15Check(
+    size_t   current_ic,
+    size_t   current_register_group,
+    uint8_t *rx_cell_voltages);
+
+static uint16_t Io_CalculatePec15(uint8_t *data, uint32_t size)
+{
+    size_t   pec15_index;
+    uint16_t pec15 = 16U;
+
+    for (size_t i = 0U; i < size; i++)
+    {
+        pec15_index = ((pec15 >> 7) ^ data[i]) & 0xFF;
+        pec15       = (uint16_t)((pec15 << 8) ^ lookup_tables.crc[pec15_index]);
+    }
+
+    // Set the LSB of the computed PEC15 to 0
+    return (uint16_t)(pec15 << 1);
+}
+
+static ExitCode Io_LTC6813_EnterReadyState(void)
+{
+    uint8_t rx_data;
+    for (size_t i = 0U; i < NUM_OF_LTC6813; i++)
+    {
+        Io_SharedSpi_Receive(ltc_6813.spi, &rx_data, 1U);
+    }
+
+    return EXIT_CODE_OK;
+}
+
+static ExitCode Io_LTC6813_StartADCConversion(void)
+{
+    uint8_t tx_cmd[NUM_OF_CMD_BYTES];
+    tx_cmd[0] = (uint8_t)(ADCV >> 8);
+    tx_cmd[1] = (uint8_t)ADCV;
+
+    uint16_t tx_cmd_pec15 =
+        Io_CalculatePec15(tx_cmd, NUM_OF_PEC15_BYTES_PER_CMD);
+    tx_cmd[2] = (uint8_t)(tx_cmd_pec15 >> 8);
+    tx_cmd[3] = (uint8_t)tx_cmd_pec15;
+
+    return (Io_SharedSpi_Transmit(ltc_6813.spi, tx_cmd, NUM_OF_CMD_BYTES) ==
+            HAL_OK)
+               ? EXIT_CODE_OK
+               : EXIT_CODE_UNIMPLEMENTED;
+}
+
+static ExitCode Io_LTC6813_PollAdcConversion(void)
+{
+    uint8_t tx_cmd[NUM_OF_CMD_BYTES];
+    uint8_t rx_data = 0xFF;
+
+    tx_cmd[0] = (uint8_t)PLADC;
+    tx_cmd[1] = (uint8_t)(PLADC >> 8);
+
+    uint16_t tx_cmd_pec15 =
+        Io_CalculatePec15(tx_cmd, NUM_OF_PEC15_BYTES_PER_CMD);
+    tx_cmd[2] = (uint8_t)(tx_cmd_pec15 >> 8);
+    tx_cmd[3] = (uint8_t)tx_cmd_pec15;
+
+    uint32_t adc_conversion_timeout_counter = 0U;
+
+    // If the data read back from the chip after a PLADC command is not equal to
+    // 0xFF, all chips on the daisy chain have finished converting cell
+    // voltages.
+    while (rx_data == 0xFF)
+    {
+        if (Io_SharedSpi_TransmitAndReceive(
+                ltc_6813.spi, tx_cmd, NUM_OF_CMD_BYTES, &rx_data, 1U) != HAL_OK)
+        {
+            return EXIT_CODE_UNIMPLEMENTED;
+        }
+
+        ++adc_conversion_timeout_counter;
+
+        // Timeout counter threshold of 10 was chosen arbitrarily.
+        if (adc_conversion_timeout_counter >= 10U)
+        {
+            return EXIT_CODE_TIMEOUT;
+        }
+    }
+
+    return EXIT_CODE_OK;
+}
+
+static void Io_LTC6813_ParseCellsAndPerformPec15Check(
+    size_t   current_ic,
+    size_t   current_register_group,
+    uint8_t *rx_cell_voltages)
+{
+    size_t cell_voltage_index = current_ic * NUM_OF_CELL_VOLTAGE_RX_BYTES;
+
+    for (size_t current_cell = 0U;
+         current_cell < NUM_OF_CELLS_PER_LTC6813_REGISTER_GROUP; current_cell++)
+    {
+        assert(
```

Comment:
@garethellis0  Not sure if this is the best way to do it but I wanted to prevent access to unitialized memory locations. If there's a better way of doing this please let me know

Reply:
I think this is fine.